lib/tests/test_revset.rs

```diff
@@ -1710,6 +1710,12 @@ fn test_evaluate_expression_difference(use_git: bool) {
     let commit4 = graph_builder.commit_with_parents(&[&commit3]);
     let commit5 = graph_builder.commit_with_parents(&[&commit2]);
 
+    // Difference from all
```

Comment:
> `~~x` isn't actually equivalent to `x` if `x` is a hidden commit

Good point. I didn't consider the case that `all()` isn't all.

Maybe I need to remove the substitution rule `all() & x` -> `x`, too. I think `author(..) & x` should include `x` even if `x` is hidden, so `Filter` expression would have to be initialized to the real `All` (or `Universe`?), and fix `all()` not to be resolved to `All`. 

> Do you think we should describe the operator that way, i.e. as "revisions that are in `all()` but not in `x`"? We could even describe it as `all() - x`? That also suggests that we should keep the `~` infix operator, I think.

Yeah, that sounds good to me. It could be defined as `~x := all() & ~x` recursively. ;)

Reply:
> Good point. I didn't consider the case that `all()` isn't all.

I can see how that is confusing when you put it like that :) I wonder if we should rename `all()` to `visible()`. I checked what Sapling does, and their model seem very similar to ours and they still call it `all()`. See https://sapling-scm.com/docs/internals/internal-difference-hg. I don't feel strongly and I'm leaning towards just leaving at as `all()` for now. If we decide to rename it to `visible()`, I think we'll also have to decide if we should also rename `heads()` to `visible_heads()`.

@quark-zju and/or @DurhamG, can you share how your users have felt about `all()` actually being only the visible commits? Have you considered renaming it?

Reply:
> I think `author(..) & x` should include `x` even if `x` is hidden

Wait, this means `author(..) & x` doesn't work without the tree substitution step. Maybe I need to think that further.

Reply:
Hmm, good question. I suppose we can check what Sapling does in that case, too. I don't have access to my desktop where Sapling is installed right now, however, because it's powered down for a move to another office (across the street).

Reply:
Actually, knowing that Sapling is based on hg, and how well you know hg's revset implementation, you can probably guess pretty well how it works there :) My guess is that Sapling does the substitution you mentioned, so `author(..) & x` becomes `x & author(..)` and does include `x`.

Reply:
Regarding hg revset, which is basically a filter from the initial `subset` (except for wdir/null),
`all()` can include hidden revisions if they're populated with `--config experimental.directaccess=true`. So rewriting the AST that way is valid.

Maybe `all()` could be extended to `:(visible_heads() | hashes_in_revsets)`?

Reply:
> Maybe `all()` could be extended to `:(visible_heads() | hashes_in_revsets)`?

That's an interesting idea. That might work. I suppose it would also need to include any commits resolved from e.g. remote branches (which are not necessarily visible).

Reply:
@martinvonz we've never had anyone comment on the fact that revsets don't usually return hidden commits. From a normal user perspective I think all() only returning visible commits matches their expectations better too.

That said, very, very few users actually delve this deep into revsets, so us hearing no complaints probably doesn't mean much.

Reply:
If thousands of your users have not been confused by it (even though most might not even have used it), then that seems like a very good indication that `all()` is good enough. Thanks, @DurhamG!

Reply:
@martinvonz so is it okay to leave the `all()` behavior a bit flaky for now?
For example, `hidden_hash & author(..)` includes `hidden_hash`, but `hidden_hash & (author(..) | whatever)` wouldn't. This is basically the same problem as `~~x` -> `x`.

Reply:
> @martinvonz so is it okay to leave the `all()` behavior a bit flaky for now?
> For example, `hidden_hash & author(..)` includes `hidden_hash`, but `hidden_hash & (author(..) | whatever)` wouldn't. This is basically the same problem as `~~x` -> `x`.

Yes, I think it's going to take a very long time until someone notices anyway, so I'm okay with that if you are.

Reply:
I think this one is resolved then. I'll add inline comment about evaluation of `RevsetExpression::All`.

Reply:
We also have a `_all()` for commits including invisible ones.

Reply:
FWIW we removed the repoview layer so there is no longer directaccess. You can read https://sapling-scm.com/docs/internals/internal-difference-hg to get a general idea about the changes.

Reply:
Yes, I saw, thanks. That's why I said your model seems very similar to ours :)

Reply:
To clarify:

* we keep track of visible heads
* we allow you to access a hidden head by specifying its commit id
* `all()` resolves to all visible commits
* `heads()` is `heads(all())`
* `descendants(x)` is all visible descendants

We don't have support for phases yet.